File: 0211_DesignAddandSearchWordsDataStructure/main.cpp
#include <string>

using namespace std;

class TrieNode {
private:
    TrieNode* next[26];
    bool isWord;
public:
    TrieNode():isWord(false), next() {}
    friend class WordDictionary;
};

class WordDictionary {
    TrieNode* root;
public:
    WordDictionary() {
        root = new TrieNode();
    }
    
    void addWord(string word) {
        TrieNode* node = root;
        for(int i = 0; i < word.size(); ++i)
        {
            if(!node->next[word[i]-'a'])
            {
                node->next[word[i]-'a'] = new TrieNode();
            }
            node = node->next[word[i]-'a'];
        }
        node->isWord = true;
    }
    
    bool search(string word) {
        return helper(word, 0, root);
    }
private:
    bool helper(string word, int startIndex, TrieNode* node)
    {
        for(int i = startIndex; i < word.size() && node; ++i)
        {
            if(word[i] == '.')
            {
                for(int j = 0; j < 26; j ++)
                {
                    if(helper(word, i+1, node->next[j])) return true;
                }
                return false;

            }
            else
            {
                node = node->next[word[i]-'a'];
            }
        }
        return node && node->isWord;
    }
};

/**
 * Your WordDictionary object will be instantiated and called as such:
 * WordDictionary* obj = new WordDictionary();
 * obj->addWord(word);
 * bool param_2 = obj->search(word);
 */

File: 0211_DesignAddandSearchWordsDataStructure/main.py
from typing import Dict, Optional

class Node:
    def __init__(self):
        self.children: Dict[str, Optional(Node)] = {}
        self.end: bool = False

class WordDictionary:
    def __init__(self):
        self.root = Node()

    def addWord(self, word: str) -> None:
        cur = self.root
        for ch in word:
            if ch in cur.children:
                cur = cur.children[ch]
            else:
                cur.children[ch] = Node()
                cur = cur.children[ch]
        cur.end = True

    def search(self, word: str) -> bool:
        def dfs(node, index):
            if index == len(word):
                return node.end
            
            if word[index] == ".":
                for child in node.children.values():
                    if dfs(child, index+1):
                        return True
                    
            if word[index] in node.children:
                return dfs(node.children[word[index]], index+1)
            
            return False
        return dfs(self.root, 0)

# Your Trie object will be instantiated and called as such:
# Your WordDictionary object will be instantiated and called as such:
obj = WordDictionary()
obj.addWord("bad")
obj.addWord("dad")
obj.addWord("mad")
param_2 = obj.search("pad")
param_2 = obj.search("bad")
param_2 = obj.search(".ad")
param_2 = obj.search("b..")